orbit_particles.c

```diff
@@ -954,3 +966,112 @@ void konestep(Config_t* cfg_ptr, int k){
   } //j
   return;
 }
+
+#ifdef __NVCC__
+__host__ __device__
+#endif
+int get_idm(Particles_t* ptcl_ptr){
+  return ptcl_ptr->idm;
+}
+
+
+void sampledep(Config_t* cfg_ptr){
+  /* config */
+  const double engn = get_engn(cfg_ptr);
+  Particles_t* ptcl_ptr = cfg_ptr->ptcl_ptr;
+
+  /* ptcl */
+  double* b = ptcl_ptr->b;
+  double* en = ptcl_ptr->en;
+  double* rho = ptcl_ptr->rho;
+  double* pot = ptcl_ptr->pot;
+  double* pol = ptcl_ptr->pol;
+  double* ptch = ptcl_ptr->ptch;
+  double* zet = ptcl_ptr->zet;
+  double* thet = ptcl_ptr->thet;
+  double* rmu = ptcl_ptr->rmu;
+  const double ekev = get_ekev(ptcl_ptr);
+
+
+  /* eqlb */
+  const double pw = get_pw(cfg_ptr->eqlb_ptr);
+
+  /* locals */
+  const char *mode = "r";
+  FILE* ifp;
+  int nheader;
+  int nlines;
+  int lineno;
+  double rdum,zdum,ptdum,edum;
+  double thetd, pold;
+  int k;
+
+  /* we need to open a file */
+  ifp = fopen(cfg_ptr->fbmdata_file, mode);
+  if (ifp == NULL) {
+    fprintf(stderr, "Can't open input file %s!\n", cfg_ptr->fbmdata_file);
+    exit(1);
+  }
+  printf("Parsing particle deposition file %s\n",  cfg_ptr->fbmdata_file);
+
+
+  /* figure out how many lines of rz data */
+  fscanf(ifp, "%d %d", &nheader, &nlines);
+  printf("sampledep found %d header lines and %d data lines\n", nheader, nlines);
+
+  /* init counters */
+  k = 0;
+  /* read in the data */
+  /* note, already read first line */
+  for(lineno=0; lineno<nlines-1; lineno++){
+    /* skip the nheader additional header lines */
+    if(lineno < nheader) fscanf(ifp, "%*[^\n]\n");
+
+    /* stay in bounds */
+    if(k >= ptcl_ptr->nprt){
+      fprintf(stderr,
+              "Error, we have more particles than nprt.  Increase nprt towards %d\n" \
+              " or, alter the code to admit min(nprt,k)...\n", nlines);
+      exit(1);
+    }
```

Comment:
See comment above. I'd replace this with a warning if nprt > nlines.
If(k >= nlines-1) -> exit from loop & reset nprt=nlines

Reply:
I'll do one check if nprt is even possible with the supplied file.  However, since the batching structure is still in place, that probably still isn't right...

I'll do a seperate check once we hit k that exits the loop (that isn't really an error so long as the particles in fbm are a random sample like we talked about).